pytket/pytket/mbqc_compile/mpattern.py

```diff
@@ -0,0 +1,434 @@
+                if self.outputs[o] == original_labels[v]:
+                    self.outputs[o] = v
+                    break
+    
+    def entangle(g: Graph) -> Circuit:
+        """
+        Creates a tket circuit which implements the edges of a zx diagram
+        via CZ gates on pairs of qubits.
+        
+        :param g:       A zx diagram.
+        :param type:    Graph
+        
+        :returns:       A pytket circuit.
+        :rtype:         Circuit
+        """
+        c = Circuit(len(g.vertices()),len(g.vertices()))
+        vlist = list(g.vertices())
+        dlist = []
+        for v in vlist:
+            dlist.append(len(g.neighbors(v)))
+        vlist = [x for _, x in sorted(zip(dlist, vlist))]
+        vlist.reverse()
+        edge_pool = set(g.edge_set())
+        finished_edge_pool = {}
+        doneround = {}
+        for v in vlist:
+            doneround[v] = False
+        while len(edge_pool)>0:
+            for vid in range(len(vlist)):
+                if not doneround[vlist[vid]]:
+                    for vid2 in range(vid+1,len(vlist)):
+                        if ((not doneround[vlist[vid]]) and (not doneround[vlist[vid2]])):
+                            if (((vlist[vid],vlist[vid2]) in edge_pool) or ((vlist[vid2],vlist[vid]) in edge_pool)):
+                                c.CZ(vlist[vid],vlist[vid2])
+                                doneround[vlist[vid]] = True
+                                doneround[vlist[vid2]] = True
+                                edge_pool -= set([(vlist[vid],vlist[vid2])])
+                                edge_pool -= set([(vlist[vid2],vlist[vid])])
+                                if vlist[vid] in finished_edge_pool.keys():
+                                    finished_edge_pool[vlist[vid]] += 1
+                                else:
+                                    finished_edge_pool[vlist[vid]] = 1
+                                if vlist[vid2] in finished_edge_pool.keys():
+                                    finished_edge_pool[vlist[vid2]] += 1
+                                else:
+                                    finished_edge_pool[vlist[vid2]] = 1
+                                break
+                        else:
+                            continue
+                else:
+                    continue
+            for key in doneround.keys():
+                doneround[key] = False
+            dlist = []
+            for v in vlist:
+                if v in finished_edge_pool.keys():
+                    dlist.append(len(g.neighbors(v))-finished_edge_pool[v])
+                else:
+                    dlist.append(len(g.neighbors(v)))
+                    finished_edge_pool[v] = 0
+            vlist = [x for _, x in sorted(zip(dlist, vlist))]
+            vlist.reverse()
+            c.add_barrier(range(len(g.vertices())), range(len(g.vertices())))
+        return c
+
+    def remove_redundant(self, g: Graph) -> None:
```

Comment:
Is this not something already supported by pyzx?

Reply:
We discussed this a while back, I'm basically merging some of the boundary vertices connected with simple edges which pyzx didn't get rid of.

Reply:
ha yeah I wrote this on 19 Jan